data.py

```diff
@@ -1,53 +1,95 @@
+        """ Return the data value in string format, if the expiry hasn't exceeded.
+            Otherwise it will return None.
+
+            Args:
+                decimals: Integer representing decimal places of the data point
+                scalar: Integer used to multiply the data value
+        """
+        if self.data_type is str:
+            return self.get()
+
+        if self.is_valid():
+            format_str = f"{{:.{decimals}f}}"
+            return format_str.format(self.value * scalar)
+        return None
+
+    def update(self, value: Any) -> None:
+        """ Update the data value and time it was updated.
+
+            Args:
+                value: Any type representing the data point of the field
+        """
+        self.value = value
+        self.time_updated = time()
+
+    def is_valid(self) -> bool:
+        """ Assess whether data is valid by checking if the valid duration
+            has exceeded. Return True if current time is less than the time
+            when data expires."""
+        return time() < self.time_updated + DataValue.DATA_EXPIRY
+
+
 class Data(ABC):
     """ A class to keep track of the most recent bike data for the overlays.
 
         Data comes into the class in the V2/V3 MQTT formats and may be accessed
         by using this class as a dictionary. This class is implemented by
         versions for specific bikes (DataV2, DataV3,...) """
 
-    data_types = {
-        # DAS data
-        "power": int,
-        "cadence": int,
-        "heartRate": int,
-        "gps": int,
-        "gps_speed": float,
-        "reed_velocity": float,
-        "reed_distance": float,
-        # Power model data
-        "rec_power": float,
-        "rec_speed": float,
-        "predicted_max_speed": float,
-        "zdist": float,
-        "plan_name": str,
-    }
-
     def __init__(self):
         # This is by no means a complete list of data fields we could track -
         # just the ones we currently think we might use on the overlays.
         self.data = {
             # DAS data
-            "power": 0,
-            "cadence": 0,
-            "heartRate": 0,
-            "gps": 0,
-            "gps_speed": 0,
-            "reed_velocity": 0,
-            "reed_distance": 0,
+            "power": DataValue(int),
+            "cadence": DataValue(int),
+            "heartRate": DataValue(int),
+            "gps": DataValue(int),
+            "gps_speed": DataValue(float),
+            "reed_velocity": DataValue(float),
+            "reed_distance": DataValue(float),
             # Power model data
-            "rec_power": 0,
-            "rec_speed": 0,
-            "predicted_max_speed": 0,
-            "zdist": 0,
-            "plan_name": "",
+            "rec_power": DataValue(float),
+            "rec_speed": DataValue(float),
+            "predicted_max_speed": DataValue(float),
+            "zdist": DataValue(float),
+            "plan_name": DataValue(str),
```

Comment:
This looks really great! Here's an extension you could do to remove some code you've actually made rather redundant. Right now, messages from the dashboard are stored separately in `self.message` and we store their received time in `self.message_received_time`. Sound familiar? It's basically what you've implemented with `DataField`. So, really we could replace all this logic by storing messages in a data field. Here's what you could do:

1. Make `DATA_EXPIRY` a member variable of `DataValue`, so that it can be set per instance
2. Make data expiry an optional parameter of `DataValue`'s constructor, defaulting to 5 but able to be set to a custom value
3. Replace all the `self.message*` stuff with a single stored instance of a `DataValue`, representing message. Give it a longer expiry timeout, say 20 seconds.
4. Update all the message related functions to just call `DataValue` methods:
    - `load_message` calls `update`
    - `has_message` calls `is_valid`
    - `get_message` calls `get_string`

If you reckon this would take too long or is beyond the scope of this PR, I'm happy for you to leave it for later and just get this PR in. If so, make a notion card and copy this info or a link to the comment in!

Reply:
Did this but I was thinking if I should just replace all the `load_message` with `update` instead, and also for the others. But since I've never used DAShboard before and don't know how to send messages onto the overlay yet, there's no way for me to test this. So I've just did the method calls

Reply:
Working brilliantly! You can send messages using only MQTT with
```bash
mosquitto_pub -t "/v3/camera/message" -m '{"message":"hello there"}'
```
although I do remember having some funky issues with quotes on windows, I don't remember which you have to use but I might leave you to figure out what needs doing.

As for running the dashboard, it's fairly straightforward - documented in the readme, but the short of it is
- Install [yarn](https://yarnpkg.com/)
- In one terminal, `cd server`, `yarn install`, `yarn start`
- In another, `cd client`, `yarn install`, `yarn start`

And you should be good to go!

Reply:
I'm happy with `load_message` still being called that. It means that things interacting with `Data`'s message stuff don't need to worry about the underlying `DataValue` implementation. That's what made this change so easy, you were able to just change how the functions worked internally, and not worry about the stuff that calls the functions at all - things just worked, without even testing!